grammars/README.md

```diff
@@ -89,3 +89,13 @@ This guide provides a brief overview. Check out the GBNF files in this directory
 ```
 ./main -m <model> --grammar-file grammars/some-grammar.gbnf -p 'Some prompt'
 ```
+
+## Troubleshooting
+
+Grammars currently have performance gotchas (see https://github.com/ggerganov/llama.cpp/issues/4218).
+
+### Efficient optional repetitions
+
+A common pattern is to allow repetitions of a pattern `x` up to N times.
+
+While semantically correct, the syntax `x? x? x?.... x?` (with N repetitions) will result in extremely slow inference. Instead, you can write `(x (x (x ... (x)?...)?)?)?` (w/ N-deep nesting)
```

Comment:
> While semantically correct, the syntax `x? x? x?.... x?` (with N repetitions) will result in extremely slow inference.

This was true prior to #6616, but I don't think it's "extremely" slow anymore.

Comparing consecutive repetitions vs. nested repetitions, there's definitely a significant speedup switching to the nested variety, but it's no longer CPU-crushingly slow to use chained ambiguity.

In my small sample test:
```
Chained ambiguity: 71 us
Chained ambiguity (grouped): 19 us
```

Compared with the previous result of:
```
Chained ambiguity: 238687 us
Chained ambiguity (grouped): 45 us
```

I think we should still encourage people to use nested ambiguity, but might just want to remove the "extremely" word here.

Reply:
The example (hidden away in a zippy) I put in https://github.com/ggerganov/llama.cpp/pull/6555#issuecomment-2051364030 still seems to explode big time with the ambiguous repetitions, it's just only visible at much higher reps now.

Reply:
Fair enough -- that's a valid benchmark data point that I was neglecting. I'm content to keep "extremely slow" in here. Carry on! 👍